packages/utils/src/testUtils.ts

```diff
@@ -53,6 +44,82 @@ export async function testEditor(spec: TestEditorSpec): Promise<void> {
+ * Also remove these from the test container.
+ */
+function _parseTextualRange(testContainer: Node): DomRangeDescription {
+    let startContainer: Node;
+    let endContainer: Node;
+    let startOffset: number;
+    let endOffset: number;
+    let direction = Direction.FORWARD;
+    function __analyseChildren(node: Node): boolean {
+        return Array.from(node.childNodes).some(child => {
+            if (child.nodeType !== Node.TEXT_NODE) {
+                return child.hasChildNodes() && __analyseChildren(child);
+            }
+            const hasOnlyRange = Array.from(child.textContent).every(char => {
+                return char === RANGE_HEAD_CHAR || char === RANGE_TAIL_CHAR;
+            });
+            const startIndex = child.textContent.indexOf(RANGE_TAIL_CHAR);
+            if (startIndex !== -1) {
+                if (hasOnlyRange) {
+                    startContainer = child.parentNode;
+                    startOffset = Array.from(child.parentNode.childNodes).indexOf(child);
+                } else {
+                    startContainer = child;
+                    startOffset = startIndex;
+                }
+                if (endContainer) {
+                    direction = Direction.BACKWARD;
+                }
+                child.textContent = child.textContent.replace(RANGE_TAIL_CHAR, '');
+            }
+            const endIndex = child.textContent.indexOf(RANGE_HEAD_CHAR);
+            if (endIndex !== -1) {
+                if (hasOnlyRange) {
+                    endContainer = child.parentNode;
+                    endOffset = Array.from(child.parentNode.childNodes).indexOf(child);
+                } else {
+                    endContainer = child;
+                    endOffset = endIndex;
+                }
+                if (startIndex > endIndex) {
+                    direction = Direction.BACKWARD;
+                    startOffset--;
+                }
+                if (child.textContent.length === 1) {
+                    child.remove();
+                }
+                child.textContent = child.textContent.replace(RANGE_HEAD_CHAR, '');
+            }
+            if (!child.textContent.length) {
+                child.remove();
+            }
+            return !!startContainer && !!endContainer;
+        });
+    }
+    if (__analyseChildren(testContainer)) {
+        return {
+            startContainer: startContainer,
+            startOffset: startOffset,
+            endContainer: endContainer,
+            endOffset: endOffset,
+            direction: direction,
+        };
+    }
+}
+function _setRange(range: DomRangeDescription): void {
+    const domRange = document.createRange();
+    domRange.setStart(range.startContainer, range.startOffset);
+    domRange.collapse(true);
+    const selection = range.startContainer.ownerDocument.getSelection();
+    selection.removeAllRanges();
+    selection.addRange(domRange);
+    selection.extend(range.endContainer, range.endOffset);
```

Comment:
Didn't we already had a function like this one in the renderer?

Reply:
We do. Should I make it public?